### ISSUE TITLE
AttributeError: 'Config' object has no attribute 'config' when packaging a python extension

### ISSUE THREAD

**Issue by s22chan** [conda/conda-build#3093]
### Actual Behavior

During packaging of a misconfigured meta.yaml for a python extension, conda-build throws an obscure AttributeError:
```python
Traceback (most recent call last):
  File "/opt/conda/envs/gk_build/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/conda_build/cli/main_build.py", line 424, in main
    execute(sys.argv[1:])
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/conda_build/cli/main_build.py", line 415, in execute
    verify=args.verify)
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/conda_build/api.py", line 200, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/conda_build/build.py", line 2209, in build_tree
    notest=notest,
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/conda_build/build.py", line 1600, in build
    built_package = bundlers[output_d.get('type', 'conda')](output_d, m, env, stats)
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/conda_build/build.py", line 847, in bundle_conda
    top_build = metadata.get_top_level_recipe_without_outputs().get('build', {}) or {}
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/conda_build/metadata.py", line 2202, in get_top_level_recipe_without_outputs
    alt_name=alt_name))
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/conda_build/metadata.py", line 1480, in _get_contents
    rendered = template.render(environment=env)
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 3, in top-level template code
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/conda_build/jinja_context.py", line 118, in load_setup_py_data
    check_call_env(args, env=get_environ(config))
  File "/opt/conda/envs/gk_build/lib/python2.7/site-packages/conda_build/environ.py", line 246, in get_dict
    prefix = m.config.host_prefix
AttributeError: 'Config' object has no attribute 'config'
```

### Expected Behavior

conda-build should throw a more helpful error message (meta.yaml should not define python in build section or something)

### Steps to Reproduce

Build a python extension with python defined in the build section.

```yaml
package:
  name: test
  version: "{{ load_setup_py_data().version }}"

source:
  path: ..

requirements:
  build:
    - {{ compiler('cxx') }}
    - python =2

  host:
    - python =2
```

This causes `conda_build/jinja_context.py:load_setup_py_data` to go into the `if os.path.isfile(config.build_python):` branch.

##### `Output of conda info`
```
     active environment : gk_build
    active env location : /opt/conda/envs/gk_build
            shell level : 2
       user config file : /root/.condarc
 populated config files : /root/.condarc
          conda version : 4.5.4
    conda-build version : not installed
         python version : 2.7.15.final.0
       base environment : /opt/conda  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/linux-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/pro/linux-64
                          https://repo.anaconda.com/pkgs/pro/noarch
                          https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /opt/conda/pkgs
                          /root/.conda/pkgs
       envs directories : /opt/conda/envs
                          /root/.conda/envs
               platform : linux-64
             user-agent : conda/4.5.4 requests/2.18.4 CPython/2.7.15 Linux/4.9.93-linuxkit-aufs debian/9 glibc/2.24
                UID:GID : 0:0
             netrc file : None
           offline mode : False
```

**Comment by piec** [comment]
Hi I also have this issue.
I only get this error with recipes that use the `compiler()` function

**Comment by msarahan** [comment]
conda build does poorly with things that pin python like that:

    - python =2

We should show warnings or errors on that.  It's a very common point of confusion.

Why do you have python as both build and host, though?  That's almost certainly not what you want.  Put it in host only.  Please read the docs on what build and host are for: https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#requirements-section

**Comment by piec** [comment]
Hi,

I'm not using a pinned python in my case, but something like that:
```
package:
  name: test
  version: "{{ load_setup_py_data().version }}"

source:
  path: ..

requirements:
  build:
    - pip
    - python
    - setuptools
    - numpy
    - cython
    - protobuf
    - {{ compiler('c') }}
  host:
    - python
    - numpy
  run:
    - python
    - numpy
```
which generate the same error

**Comment by msarahan** [comment]
> Why do you have python as both build and host, though? That's almost certainly not what you want. Put it in host only. Please read the docs on what build and host are for: https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#requirements-section

**Comment by piec** [comment]
OK I removed `python` from `build` -- I still have the same error though

---
I have read the documentation but I don't fully understand why conda has a build / host distinction. Other cross-compilation environment that I've used only differentiate between build-time and run-time

**Comment by msarahan** [comment]
If you only removed python from build, your other dependencies will still bring it in.

Cross-compilation always involves 3 environments: the native one (the platform you're building on), the target one (the one you're building for) and the runtime one (same platform as the target one).  Any other cross-compilation environment is hiding this complexity from you with some implicit construct.  Build tools for the native platform must be used, because non-native build tools won't run.  On the other hand, you need target-platform libraries to link against.

From the docs:

> TL;DR: If you use the new {{ compiler() }} jinja2 to utilize our new compilers, you also must move anything that is not strictly a build tool into your host dependencies. This includes python, python libraries, and any shared libraries that you need to link against in your build. Examples of build tools include any {{ compiler() }}, make, autoconf, perl (for running scripts, not installing perl software), python (for running scripts, not for installing software).

Is that not clear?

**Comment by mingwandroid** [comment]
> I have read the documentation but I don't fully understand why conda has a build / host distinction. Other cross-compilation environment that I've used only differentiate between build-time and run-time

And what do you think is different about these two schemes? "build == build-time", "host == run-time". We also allow the user to specify their own `run` dependencies (though due to `run_exports` there is not much need).

Our separation scheme provides flexibility to use static libraries from other packages in your packages too (use `ignore_run_exports`).

**Comment by s22chan** [comment]
@msarahan IMHO, any TLDR should be at the beginning, instead of the end of a wall of text.

**Comment by mingwandroid** [comment]
The TL;DR was in a quote, but good try!

**Comment by s22chan** [comment]
@mingwandroid I was referring to the original readthedocs :)

**Comment by msarahan** [comment]
If your answer to our help is to nitpick our docs, I'm afraid you may get less than polite responses.  As Ray says, please consider a PR.

**Comment by piec** [comment]
@msarahan @mingwandroid ok it's clearer now, I have a few questions:

About the build process, what are the "PREFIX environment"s (host prefix, build prefix)? I guess they are conda environments created during the build, in `conda-bld/...`?
I tried to make `conda-build` leave these environments around after the build but I didn't succeed (with `--no-remove-work-dir` or `--keep-old-work`)

Is this correct:
* build: build-time dependency, installed in the build prefix?
* host: build-time dependency, installed in the host prefix?
* run: not related to the build itself - run-time dependencies, when a user installs the package 

---
> Executables that don’t exist in the host prefix should be found in the build prefix
How can executable from the host prefix be executed during build time?
(For example when cross-compiling on an intel 64b linux machine for an arm 64b linux machine)

---
Do you have an example of "cross-capable" recipe?

---
What should cython be in my example: build / host?
For more context cython is used to cythonize a few .pyx files to .c, which are then compiled to dynamic libraries (.so)
If I follow the rules above it should be in build in my opinion, because the cython build tool runs on the builder os.
But cython depends on python which brings python to the build array which conda-build doesn't like.

Cheers

**Comment by mingwandroid** [comment]
> How can executable from the host prefix be executed during build time?
(For example when cross-compiling on an intel 64b linux machine for an arm 64b linux machine)

You need to use something like qemu and binfmt_misc for that. I've done this before.

**Comment by mingwandroid** [comment]
Cython should be in `host`, although it's a build tool we don't currently support python-based cross compilation where `python` itself would be in build, but `python` and it packages would also be in host, so that forces it into `host`. I've done some work for python-based cross but it's not ready and I don't believe the ecosystem is setup to handle it either (many setup.py scripts will ignore our attempts to cross build things).

**Comment by mingwandroid** [comment]
Do you mean is there a way to replace the build prefixes automatically in text and binaries? Host prefixes are automatically replaced but at present you need to use `sed` on the builds ones manually or in some cases strip off full paths (e.g. `export M4=m4` or `export CC=$(basename $}CC})`.

**Comment by piec** [comment]
I think I didn't use the correct term, I meant conda environments of the build.
From what I gather conda-build creates clean conda environment for the lifetime of the build is there a way to keep these directories in order to inspect what happened afterwards? Maybe activate them manually?

**Comment by mingwandroid** [comment]
`--keep-old-work` and `--dirty` can be used, alternatively, drop an `exit 1` in your script and `conda-build` will not remove anything.
